shared-bindings/displayio/Display.c

```diff
@@ -356,16 +373,95 @@ const mp_obj_property_t displayio_display_bus_obj = {
 };
 
 
+#include "py/objarray.h"
+mp_obj_array_t *array_new(char typecode, size_t n);
+mp_obj_t array_subscr(mp_obj_t self_in, mp_obj_t index_in, mp_obj_t value);
+
+//|   .. method:: fill_row(y, buffer)
+//|
+//|     Extract the pixels fro a single row
+//|
+//|     :param int y: The top edge of the area
+//|     :param bytearray buffer: The buffer in which to place the pixel data
+STATIC mp_obj_t displayio_display_obj_fill_row(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+  enum { ARG_y, ARG_buffer };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_y, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+        { MP_QSTR_buffer, MP_ARG_OBJ | MP_ARG_KW_ONLY, {} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    displayio_display_obj_t *self = native_display(pos_args[0]);
+    mp_int_t y = args[ARG_y].u_int;
+    mp_obj_array_t *result = (mp_obj_array_t *)(args[ARG_buffer].u_obj);
+
+    if (result->typecode != BYTEARRAY_TYPECODE) {
+      mp_raise_ValueError(translate("Buffer is not a bytearray."));
+    }
+    if (self->colorspace.depth != 16) {
+      mp_raise_ValueError(translate("Display must have a 16 bit colorspace."));
+    }
+
+    displayio_area_t area = {
+      .x1 = 0,
+      .y1 = y,
+      .x2 = self->width,
+      .y2 = y + 1
+    };
+    uint8_t pixels_per_word = (sizeof(uint32_t) * 8) / self->colorspace.depth;
+    uint16_t buffer_size = self->width / pixels_per_word;
+    uint16_t pixels_per_buffer = displayio_area_size(&area);
+    if (pixels_per_buffer % pixels_per_word) {
+      buffer_size += 1;
+    }
+
+    // Allocated and shared as a uint32_t array so the compiler knows the
+    // alignment everywhere.
+    uint32_t buffer[buffer_size];
+    volatile uint32_t mask_length = (pixels_per_buffer / 32) + 1;
+    uint32_t mask[mask_length];
+
+    for (uint16_t k = 0; k < mask_length; k++) {
+      mask[k] = 0x00000000;
+    }
+    for (uint16_t k = 0; k < buffer_size; k++) {
+      buffer[k] = 0x00000000;
+    }
+
+    displayio_display_fill_area(self, &area, mask, buffer);
+
+    if ((result->len + result->free) >= (buffer_size * 4)) {
```

Comment:
Please do this check before fill_area and then pass the given buffer straight into fill_area. That will simplify this implementation further.